File: src/execution/runtimeRecoveryEngine.ts
/**
 * Converted from Python: core/execution/runtime_recovery_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function recoverRuntimeExecution(failed_actions: any = null, checkpoint: any = null, interrupted_workflows: any = null): any {
  var failed: any = [...py.iter(py.or2(failed_actions, () => ([])))];
  checkpoint = py.or2(checkpoint, () => ({}));
  var workflows: any = [...py.iter(py.or2(interrupted_workflows, () => ([])))];
  var recovered_actions: any = py.enumerate(py.sorted(failed, {key: ((item: any) => py.toStr(py.get(item, "id", ""))) as (item: any) => any})).map(([index, action]: any) => ({...(action), "recovered": true, "replay_index": index}));
  return {"recovered_actions": recovered_actions, "checkpoint_restored": py.truthy(checkpoint), "workflows_resumed": py.len(workflows), "sync_divergence_resolved": true, "replay_safe": true, "bounded": true};
}

File: lib/src/execution/runtime_recovery_engine.dart
/// Port of core/execution/runtime_recovery_engine.recover_runtime_execution.
Map<String, dynamic> recoverRuntimeExecution({
  List<Map<String, dynamic>>? failedActions,
  Map<String, dynamic>? checkpoint,
  List<dynamic>? interruptedWorkflows,
}) {
  final List<Map<String, dynamic>> failed = <Map<String, dynamic>>[
    ...(failedActions ?? <Map<String, dynamic>>[])
  ];
  final Map<String, dynamic> cp = checkpoint ?? <String, dynamic>{};
  final List<dynamic> workflows = <dynamic>[
    ...(interruptedWorkflows ?? <dynamic>[])
  ];

  final List<Map<String, dynamic>> sortedFailed = <Map<String, dynamic>>[
    ...failed
  ]..sort((Map<String, dynamic> a, Map<String, dynamic> b) =>
      '${a['id'] ?? ''}'.compareTo('${b['id'] ?? ''}'));

  final List<Map<String, dynamic>> recoveredActions = <Map<String, dynamic>>[];
  for (int index = 0; index < sortedFailed.length; index++) {
    recoveredActions.add(<String, dynamic>{
      ...sortedFailed[index],
      'recovered': true,
      'replay_index': index,
    });
  }

  return <String, dynamic>{
    'recovered_actions': recoveredActions,
    'checkpoint_restored': cp.isNotEmpty,
    'workflows_resumed': workflows.length,
    'sync_divergence_resolved': true,
    'replay_safe': true,
    'bounded': true,
  };
}
